openlibrary/core/imports.py

```diff
@@ -33,56 +33,59 @@ def load_items(self, filename):
         items = [line.strip() for line in open(filename) if line.strip()]
         self.add_items(items)
 
-    def dedupe_ia_items(self, items):
-        already_present = [
+    def dedupe_items(self, items):
+        ia_ids = [item.get('ia_id') for item in items if item.get('ia_id')]
+        already_present = set([
             row.ia_id for row in db.query(
-                "SELECT ia_id FROM import_item WHERE ia_id IN $items",
+                "SELECT ia_id FROM import_item WHERE ia_id IN $ia_ids",
                 vars=locals()
             )
-        ]
+        ])
         # ignore already present
         logger.info(
             "batch %s: %d items are already present, ignoring...",
             self.name,
             len(already_present)
         )
-        return list(set(items) - set(already_present))
+        # Those unique items whose ia_id's aren't already present
+        return [
+            item for item in items
+            if item.get('ia_id') not in already_present
+        ]
 
-    def add_items(self, items, ia_items=True):
+    def normalize_items(self, items):
+        return [{
+            'batch_id': self.id,
+            'ia_id': item.get('ia_id') or item,
+            'data': json.dumps(
+                item.get('data'),
+                sort_keys=True
+            ) if item.get('data') else None
+        } for item in items]
+
+    def add_items(self, items):
```

Comment:
This is the best place to start the code review.

`items` used to be a list of either `ia_id` strings or `["ia_id1", "ia_id2", ..., "ia_idn"]` or a list of book records dicts `[{"data": 1}, ...]`.

Now, items is a list of dicts whose keys are `ia_id` (required) and `data` (optional).

`normalize_items` marshals this data into what we want in the database.

Now, `dedupe_items` is **always** run, both for IA & partner imports based on the `ia_id` which could be a partner-type (e.g. `isbn:0123456789`) or legacy Archive.org `ia_id` (e.g. `goody`)